miner/worker.go

```diff
@@ -551,64 +551,75 @@ func (w *worker) taskLoop() {
 func (w *worker) resultLoop() {
 	for {
 		select {
-		case block := <-w.resultCh:
-			// Short circuit when receiving empty result.
-			if block == nil {
+		case result := <-w.resultCh:
+			// Short circuit when receiving empty result or receiving duplicate result caused by resubmitting.
+			if result == nil || result.Block == nil || w.chain.HasBlock(result.Block.Hash(), result.Block.NumberU64()) {
 				continue
 			}
-			// Short circuit when receiving duplicate result caused by resubmitting.
-			if w.chain.HasBlock(block.Hash(), block.NumberU64()) {
-				continue
-			}
-			var (
-				sealhash = w.engine.SealHash(block.Header())
-				hash     = block.Hash()
-			)
-			w.pendingMu.RLock()
-			task, exist := w.pendingTasks[sealhash]
-			w.pendingMu.RUnlock()
-			if !exist {
-				log.Error("Block found but no relative pending task", "number", block.Number(), "sealhash", sealhash, "hash", hash)
-				continue
-			}
-			// Different block could share same sealhash, deep copy here to prevent write-write conflict.
-			var (
-				receipts = make([]*types.Receipt, len(task.receipts))
-				logs     []*types.Log
-			)
-			for i, receipt := range task.receipts {
-				// add block location fields
-				receipt.BlockHash = hash
-				receipt.BlockNumber = block.Number()
-				receipt.TransactionIndex = uint(i)
-
-				receipts[i] = new(types.Receipt)
-				*receipts[i] = *receipt
-				// Update the block hash in all logs since it is now available and not when the
-				// receipt/log of individual transactions were created.
-				for _, log := range receipt.Logs {
-					log.BlockHash = hash
-					// Handle block finalization receipt
-					if (log.TxHash == common.Hash{}) {
-						log.TxHash = hash
+
+			if result.IsProposer {
```

Comment:
If my understanding of the code is correct, this means that the proposer must deep copy the block submitted through the seal function, but other validators sending in the result through the commit function don't have to deep copy the result. Could you add a comment briefly explaining why this is?